enhancements/optional-manifest.md

```diff
@@ -0,0 +1,357 @@
+ServiceMonitor and VerticalPodAutoscaler are among the supported resources. They allow the operator authors to provide more advanced features (observability, autoscaling). These features are however not part of the core functionalities of most operators. As such the operator author may want to leverage them but not enforce them so that the operator bundle can be installed on clusters where they are not available. As these resources are not shipped with Kubernetes it is a significant drawback. It either holds operator authors of leveraging them or forces them to have additional logic directly in the operator for managing their creation.
+
+### Goals
+
+- Allow the specification of optional resources
+- Avoid InstallPlan failure when optional resources cannot be created due to the absence of their API
+- Keep the current behaviour for resources that are not specified as optional
+
+### Non-Goals
+
+- Although there is a relation between this enhancement proposal and optional dependencies it does not intend to cover the full scope of the later. It can however be seen as an enabler for it and aim at extending OLM in a way that is consumable by such an enhancement.
+
+## Proposal
+
+A resource can be marked as optional by leveraging [Explicit Bundle Properties](https://github.com/operator-framework/enhancements/blob/master/enhancements/properties.md). Therefore a new file can be added to the bundle metadata directory. The file itself can have any name, but it has to be in the top-level metadata directory of the bundle and to consist of YAML with a top-level `properties` element containing an array of type/value tuples. There should be a single file containing a top-level `properties` element.
+
+The property type for optional manifest is `olm.manifests.optional`. Its value field is a `files` structure, which contains a list of optional files.
+
+Example:
+
+~~~
+properties:
+- type: 'olm.manifests.optional'
+  value:
+    files:
+    - verticalpodautoscaler.yaml
+    - servicemonitor.yaml
+~~~
+
+If a file listed in the property file does not exist in the bundle manifest directory it will be ignored without further hints.
+
+It does not make sense to list a resource core to OLM (ClusterServiceVersion) or resources shipped with Kubernetes under `olm.manifests.optional` as the availability of the API is a prerequisite for OLM to work. This means that no extra check will be implemented for the following types:
+
+- ClusterServiceVersions
+- Secrets
+- ServiceAccounts
+- Roles
+- RoleBindings
+- ClusterRoles
+- ClusterRoleBindings
+- Services
+- ConfigMaps
+
+This has the advantage of limiting the scope of the change. There is no specific logic with manifests of unknown types (types not in the above list). Should the list of allowed resources get extended possibly to any resource type no further amendment should be required to this enhancement.
```

Comment:
Does it make sense for us to add validation to the bundle validator to ensure that this property isn't being set on types that it cannot support?

Reply:
Could probably pretty easily add some validation to this property here https://github.com/operator-framework/api/blob/master/pkg/validation/internal/bundle.go#L28

Reply:
It is possible but in my opinion unnecessary. Let's say that the operator author add a manifest with a service definition to the optional list. The expressed intent is that if the service cannot be created due to the missing API the InstallPlan would still continue and possibly succeed. The thing is that the service API CANNOT be missing. It comes with any Kubernetes installation. It is not based on a CRD. It is the same for all the excluded resources except the CSV. Today if the CSV API is missing OLM cannot install an operator at all.

Reply:
Should we assume that the only reason for marking a resource optional is because the API is potentially not available on the cluster? What if the resource is available, but at a specific version that does not include an optional field that is specified by the manifest in the bundle? Should we force fail there if the resource is specified as optional?

That being said, I think you're right in that it does seem reasonable to me that our API does not concern itself with specific resource availability at packaging time, because there's always going to be compatibility concerns. I just wonder if that means we should be even more generic in our statement, and rather than specify a set of resources that OLM will always apply and ignore this annotation on. Instead say that "any API that is available on the cluster will be applied as part of installation"

Reply:
That's a good point. I think that we could differentiate errors that are due to a specific cluster configuration. They should not fail the InstallPlan. Here is a proposal going through[ the list of possible errors](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L729-L873)


StatusReasonUnauthorized and StatusReasonForbidden
Due to access right configuration => Warning in the logs and go on with the InstallPlan
rational: It is not given that the cluster admin wants to allocate more rights.

StatusReasonNotFound
API not available => Warning in the logs and go on with the InstallPlan

StatusReasonAlreadyExists
resource already exists, unlikely with create except in case of a cleansing issue => Warning in the logs and go on with the InstallPlan

StatusReasonConflict
Update failure, unlikely with create => Warning in the logs and error returned

StatusReasonGone shares the same error code as StatusReasonExpired
It should be a transcient issue: "typically associated with watches that can't be serviced" => Warning in the logs and error returned so that OLM can try again

StatusReasonInvalid
invalid data provided as part of the request.
This would match the scenario where a field is unsupported by the on-cluster version.
I would tend to think that it is easier for the cluster admin to deal with the situation if the installation goes on.
2 scenarios:
- The cluster API cannot get upgraded to the desired version
 + If we go on with the installation: the resource being specified as optional it works as intended by the operator author.
 + If we stop the installation: the cluster-admin is blocked and the operator cannot get installed
- The cluster API can get upgraded to the desired version
 + If we go on with the installation. The operator is installed with limited capabilities. At some point, possibly not right away the user will notice it. Looking at the InstallPlan it is possible to see that the resource was not created. The recovery workflow should be for the cluster-admin to try to create the resource manually. This will indicate that the versions are not compatible (it needs a bit of knowledge). Alternatively if the cluster-admin checks the logs it should see a warning where we provide additional information. It can then upgrade the cluster API and create the resource manually.
 + If we stop the installation. The cluster-admin should check the logs, see a warning where we provide additional information. It can then upgrade the cluster API and retry the installation.

StatusReasonServerTimeout, StatusReasonTimeout & StatusReasonTooManyRequests
They should be transcient issues, at least not directly related with the resource being created. => Warning in the logs and error returned so that OLM can try again

StatusReasonBadRequest and StatusReasonMethodNotAllowed
mean that the request itself was invalid, because the request doesn't make any sense, for example deleting a read-only object. API calls that return BadRequest or StatusReasonMethodNotAllowed can never succeed. => Warning in the logs and error returned so that OLM fails the InstallPlan

StatusReasonNotAcceptable
means that the accept types indicated by the client were not acceptable to the server - for instance, attempting to receive protobuf for a resource that supports only json and yaml.
It may be specific to a cluster version. It seems to happen rarely. => Warning in the logs and go on with the InstallPlan

StatusReasonRequestEntityTooLarge
This should not be cluster specific. => Warning in the logs and error returned so that OLM fails the InstallPlan

StatusReasonUnsupportedMediaType
same as StatusReasonNotAcceptable in the other direction

StatusReasonInternalError
we cannot deduce anything from that => Warning in the logs and error returned so that OLM can try again

StatusReasonExpired
unlikely to happen with a create.  => Warning in the logs and error returned so that OLM can try again

StatusReasonServiceUnavailable
Retrying the request may succeed. => Warning in the logs and error returned so that OLM can try again